nbclient/client.py

```diff
@@ -1033,6 +1051,104 @@ def handle_comm_msg(
             if comm_id in self.comm_objects:
                 self.comm_objects[comm_id].handle_msg(msg)
 
+    async def async_execute_expressions(self, cell, cell_index: int, expressions: t.List[str]) -> t.Dict[str, Any]:
+        parent_msg_id = await ensure_async(
+            self.kc.execute(
+                '',
+                silent=True,
+                user_expressions={f"{MD_EXPRESSIONS_PREFIX}-{i}": expr for i, expr in enumerate(expressions)},
+            )
+        )
+        task_poll_kernel_alive = asyncio.ensure_future(
+            self._async_poll_kernel_alive()
+        )
+        task_poll_expr_msg = asyncio.ensure_future(
+            self._async_poll_expr_msg(parent_msg_id)
+        )
+        exec_timeout = None
+        self.task_poll_for_reply = asyncio.ensure_future(
+            self._async_poll_for_expr_reply(
+                parent_msg_id, cell, exec_timeout, task_poll_expr_msg, task_poll_kernel_alive
+            )
+        )
+        try:
+            exec_reply = await self.task_poll_for_reply
+        except asyncio.CancelledError:
+            # can only be cancelled by task_poll_kernel_alive when the kernel is dead
+            task_poll_expr_msg.cancel()
+            raise DeadKernelError("Kernel died")
+        except Exception as e:
+            # Best effort to cancel request if it hasn't been resolved
+            try:
+                # Check if the task_poll_output is doing the raising for us
+                if not isinstance(e, CellControlSignal):
+                    task_poll_expr_msg.cancel()
+            finally:
+                raise
+        self._check_raise_for_error(cell, exec_reply)
+        attachments = {key: val["data"] if "data" in val else {"traceback": "\n".join(val["traceback"])} for key, val in exec_reply["content"]["user_expressions"].items()}
```

Comment:
Over on `jupyterlab-imarkdown` [I'm playing around with](https://github.com/agoose77/jupyterlab-imarkdown/blob/ed6d3dfe6b6edd5947ee824dbf131bae6fe70a1b/src/kernel.ts#L51-L65 ) keeping the kernel-output itself as [two different MIME types](https://github.com/agoose77/jupyterlab-imarkdown/blob/ed6d3dfe6b6edd5947ee824dbf131bae6fe70a1b/src/attachment.ts#L3-L4), i.e. the "attachment" is just one of two possible MIME types.

My thought process here is that we keep the structural information, and do the formatting [as late as possible](https://github.com/agoose77/jupyterlab-imarkdown/blob/master/src/cell.ts#L60-L83). The `status` field probably shouldn't be stored, as it's already encoded in the MIME type.

How do you feel about doing something like this?

Reply:
Heya 👋🏼 well my feeling is that really we use have a new cell type with a new json schema, that would be a sort of hybrid between code and markdown; e.g. it should definitely have an execution number and then this data should have its own “expressions” key, where yeh we would have a more 1-to-1 mapping with the data returned by the kernel

Reply:
Having these special mime types indeed works, but I feel is a bit “hacky” for core jupyter

Reply:
so it would be along the lines of:

```json
  {
   "cell_type": "imarkdown",
   "execution_count": 1,
   "metadata": {},
   "expressions": [
      {
        "output_type": "display",
        "data": {
          "text/plain": "1"
        }
      },
      {
        "output_type": "error",
        "traceback": ["..."]
      },
   ],
   "source": [
    "{{ a }}\n"
   ]
  }
```